scrapy/core/http2/protocol.py

```diff
@@ -100,31 +104,34 @@ def connectionMade(self):
         """Called by Twisted when the connection is established. We can start
         sending some data now: we should open with the connection preamble.
         """
-        LOGGER.info("Connection made to {}".format(self.transport))
+        self.destination = self.transport.connector.getDestination()
+        LOGGER.info('Connection made to {}'.format(self.destination))
+
         self.conn.initiate_connection()
         self._write_to_transport()
-
         self.is_connection_made = True
 
-        # Initiate all pending requests
-        for stream in self._pending_request_stream_pool:
-            assert isinstance(stream, Stream)
-            stream.initiate_request()
-
-        self._pending_request_stream_pool.clear()
+        self._send_pending_requests()
 
     def dataReceived(self, data):
         events = self.conn.receive_data(data)
         self._handle_events(events)
         self._write_to_transport()
 
     def connectionLost(self, reason=connectionDone):
-
         """Called by Twisted when the transport connection is lost.
+        No need to write anything to transport here.
         """
-        for stream_id in self.streams.keys():
-            # TODO: Close each Stream instance in a clean manner
-            self.conn.end_stream(stream_id)
+        # Pop all streams which were pending and were not yet started
+        for stream_id in list(self.streams):
+            try:
+                self.streams[stream_id].lost_connection()
+            except KeyError:
+                pass
```

Comment:
It looks like you could use `self.streams.values()` instead.

I’m not sure why `KeyError` is needed, unless `list_connection()` may raise it.

Reply:
>  looks like you could use `self.streams.values()` instead.

Yup I tried using `self.stream.values()` at first but that gave me "RuntimeError: Dictionary changed size during iteration" which was because of `lost_connection()` which calls `self._stream_close_cb` which pops the element from the dictionary. I then tried [this solution](https://stackoverflow.com/questions/11941817/how-to-avoid-runtimeerror-dictionary-changed-size-during-iteration-error) which worked. 

> I’m not sure why `KeyError` is needed, unless `list_connection()` may raise it.

Just removed it and tested -- Works 🙃